input.go

```diff
@@ -0,0 +1,19 @@
+package shaping
+
+import (
+	"github.com/go-text/di"
+	"github.com/go-text/font"
+	"golang.org/x/image/math/fixed"
+)
+
+type Input interface {
+	// Text returns the characters to be shaped.
+	Text() []rune
+	// Direction returns the directionality of the text.
+	Direction() di.Direction
```

Comment:
Text script and language are also needed for proper shaping (script can be inferred from text, but it usually needs the full paragraph to handle special cases, so should be done at a higher level), language is usually specified by the user or guessed through some other means (e.g. locale) since determining text language can’t be reliability automated for the purpose of text layout.

Reply:
Agreed, thanks (as an example, I use [this package](https://pkg.go.dev/github.com/benoitkugler/textlayout@v0.0.1/language) for this.)

Note however that it depends on the scope of the package, which I think is still not well defined.
If we aim a laying out a whole paragraph, the `Input` type would probably need to provide the language for each _segment_ of the text (and the scripts would be inferred from the whole paragraph). Same question for the fonts.

So I think we have to find a way to specify information for segments (or portions) of text. I think Pango uses [attributes](https://developer.gnome.org/pango/1.46/pango-Text-Attributes.html#PangoAttribute) for this, each of them specifying one value for one attribute for one portion of text.